java-util/src/main/java/io/druid/java/util/http/client/pool/ChannelResourceFactory.java

```diff
@@ -109,6 +112,28 @@ public ChannelFuture generate(final String hostname)
 
       final ChannelPipeline pipeline = connectFuture.getChannel().getPipeline();
       pipeline.addFirst("ssl", sslHandler);
+      pipeline.addLast("connectionErrorHandler", new SimpleChannelUpstreamHandler()
+      {
+        private final Logger LOGGER = new Logger(SimpleChannelUpstreamHandler.class);
```

Comment:
It looks that we can set exceptions for `handshakeFuture` rather than logging errors and this would be better because exceptions can be propagated to the callers properly.

Reply:
Done.

Reply:
I'm seeing this code.

```java
      pipeline.addLast("connectionErrorHandler", new SimpleChannelUpstreamHandler()
      {
        @Override
        public void exceptionCaught(ChannelHandlerContext ctx, ExceptionEvent e)
        {
          final Channel channel = ctx.getChannel();
          if (channel == null) {
            // For the case where this pipeline is not attached yet.
            return;
          }
          if (channel.isOpen()) {
            channel.close();
          }
        }
      });
```

Looks like the exception is not properly set for `handshakeFuture`. What I meant is like below:

```java
      final ChannelFuture handshakeFuture = Channels.future(connectFuture.getChannel());

      pipeline.addLast("connectionErrorHandler", new SimpleChannelUpstreamHandler()
      {
        @Override
        public void exceptionCaught(ChannelHandlerContext ctx, ExceptionEvent e)
        {
          final Channel channel = ctx.getChannel();
          if (channel == null) {
            // For the case where this pipeline is not attached yet.
            handshakeFuture. setFailure(new ChannelException("Channel is null. The context name is ..."));
          }
          else {
            handshakeFuture.setFailure(e.getCause());
            if (channel.isOpen()) {
              channel.close();
            }
          }
        }
      });
```